In [1]:
import numpy as np
import pandas as pd


In [4]:
movies = pd.read_csv(r"C:\Users\haric\Desktop\ML projects\movie recomendation system\tmdb_5000_movies.csv")
credits = pd.read_csv(r"C:\Users\haric\Desktop\ML projects\movie recomendation system\tmdb_5000_credits.csv")

In [18]:
movies.head(1)


budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  ...  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...   

                                              crew_y movie_id_x  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...      19995   

                                              cast_x  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                              crew_x movie_id_y  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...      19995   

                                              cast_y  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                              crew_y movie_id  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...    19995   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  

[1 rows x 35 columns]

In [17]:
movies.shape

(4989, 35)

In [8]:
credits.head(1)


movie_id   title                                               cast  \
0     19995  Avatar  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [16]:
credits.shape

(4803, 4)

LETS MERGE THE DFS BY COLOUMN TITLE WHY WE NEED 2 DFS:-)

In [13]:
movies = movies.merge(credits, on = "title")
movies.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  ...  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...   

                                              crew_y movie_id_x  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...      19995   

                                              cast_x  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                              crew_x movie_id_y  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...      19995   

                                              cast_y  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                              crew_y movie_id  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...    19995   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  

[1 rows x 35 columns]

In [15]:
movies.shape

(4989, 35)

In [19]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4989 entries, 0 to 4988
Data columns (total 35 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4989 non-null   int64  
 1   genres                4989 non-null   object 
 2   homepage              1743 non-null   object 
 3   id                    4989 non-null   int64  
 4   keywords              4989 non-null   object 
 5   original_language     4989 non-null   object 
 6   original_title        4989 non-null   object 
 7   overview              4986 non-null   object 
 8   popularity            4989 non-null   float64
 9   production_companies  4989 non-null   object 
 10  production_countries  4989 non-null   object 
 11  release_date          4988 non-null   object 
 12  revenue               4989 non-null   int64  
 13  runtime               4987 non-null   float64
 14  spoken_languages      4989 non-null   object 
 15  status               

In [20]:
movies = movies[['movie_id','title','overview','genres','keywords','cast','crew']]

In [21]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

Now i will combine overview and genres and keywods and cast and crew into one and name it as keyword so that every movie has a keyword

for doing the above step we have to preprocess the data which includes the data cleaning and all the stuff

In [27]:
#The is null function in Python is used to check if a value is None. 
# It returns True if the value is None and False otherwise.
movies.isnull().sum()

movie_id    0
title       0
overview    0
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [26]:
movies.dropna(inplace=True)

In [33]:
#CHcking whether we have any duplicate rows or not
movies.duplicated().sum()

0

In [32]:
movies.drop_duplicates(inplace=True)

IN GENERS WE HAVE A DICTIONARY AS WE CAN SEE BELOW BUT WE NEED LIST WHICH CONATAINS ALL THE GENERS OF THE MOVIE

In [35]:
movies.iloc[1].genres

'[{"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 28, "name": "Action"}]'

HERE BUT THE THING IS ALL THE VALUES WHICH WE SEE ABOVE IS IN STRING FORMAT IF WE CINVERT THAT INTO LIST THEN IT WOULD BE VERY HELPFUL AND ALSO IN THE NEXT STEP WHAT WE WILL DO IS WE WILL CONVERT THE STRING INTO LIST AND  FIX IT for that we will import module called ast and we will use the function literal_eval in it

In this example, the literal_eval function is used to safely evaluate the code string, which contains a simple Python expression. The result of this evaluation is a Python list,

In [41]:
import ast
def converter(obj):
    list = [ ]
    for i in ast.literal_eval(obj):
        list.append(i['name'])
    return list

In [43]:
movies["genres"] = movies["genres"].apply(converter)

In [47]:
#movies["keywords"] = movies["keywords"].apply(converter)
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

for cast it is entirely different have a look into belows output
it has lot of dictionaries in it and we need to take only name from the dictionary and also we need to loop though only 3 dictionaries because they are top three cats of the movies and for doing that we will  write a function


In [48]:
movies["cast"].iloc[0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In [49]:
def converter3(obj):
    list = [ ]
    counter = 0
    for i in ast.literal_eval(obj):
        if counter!=3:
            list.append(i['name'])
        else:
            break
    return list

In [50]:
movies["cast"] = movies["cast"].apply(converter3)

In [51]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                                cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weave...   
1  [Johnny Depp, Orlando Bloom, Keira Knightley, ...   
2  [Daniel Craig, Christoph Waltz, Léa Seydoux, R...   
3  [Christian Bale, Michael Caine, Gary Oldman, A...   
4  [Taylor Kitsch, Lynn Collins, Samantha Morton,...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [52]:
movies["crew"].iloc[0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

In [54]:
def fetch_director(obj):
    list = [ ]
    for i in ast.literal_eval(obj):
        if i["job"] == "Director":
            list.append(i['name'])
            break        
    return list

In [55]:
movies["crew"]= movies["crew"].apply(fetch_director)

In [56]:
movies["crew"][0]

['James Cameron']

In [57]:
    movies["overview"] = movies["overview"].apply(lambda x : x.split())

In [58]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                                cast                 crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weave...      [James Cameron]  
1  [Johnny Depp, Orlando Bloom, Keira Knightley, ...     [Gore Verbinski]  
2  [Daniel Craig, Christoph Waltz, Léa Seydoux, R...         [Sam Mendes]  
3  [Christian Bale, Michael Caine, Gary Oldman, A...  [Christopher Nolan]  
4  [Taylor Kitsch, Lynn Collins, Samantha Morton,...     [Andrew Stanton]

In [61]:
movies["genres"] = movies["genres"].apply(lambda x :[i.replace(" ","") for i in x])
movies["keywords"] = movies["keywords"].apply(lambda x :[i.replace(" ","") for i in x])
movies["cast"] = movies["cast"].apply(lambda x :[i.replace(" ","") for i in x])
movies["crew"] = movies["crew"].apply(lambda x :[i.replace(" ","") for i in x])

In [62]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                                cast                crew  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver, ...      [JamesCameron]  
1  [JohnnyDepp, OrlandoBloom, KeiraKnightley, Ste...     [GoreVerbinski]  
2  [DanielCraig, ChristophWaltz, LéaSeydoux, Ralp...         [SamMendes]  
3  [ChristianBale, MichaelCaine, GaryOldman, Anne...  [ChristopherNolan]  
4  [TaylorKitsch, LynnCollins, SamanthaMorton, Wi...     [AndrewStanton]

In [63]:
movies["tags"] = movies["overview"] + movies["genres"] + movies["keywords"] + movies["cast"] + movies["crew"]

In [64]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                                cast                crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver, ...      [JamesCameron]   
1  [JohnnyDepp, OrlandoBloom, KeiraKnightley, Ste...     [GoreVerbinski]   
2  [DanielCraig, ChristophWaltz, LéaSeydoux, Ralp...         [SamMendes]   
3  [ChristianBale, MichaelCaine, GaryOldman, Anne...  [ChristopherNolan]   
4  [TaylorKitsch, LynnCollins, SamanthaMorton, Wi...     [AndrewStanton]   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [65]:
new_df = movies[["movie_id","title","tags"]]

In [66]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [67]:
new_df['tags'][0]

['In',
 'the',
 '22nd',
 'century,',
 'a',
 'paraplegic',
 'Marine',
 'is',
 'dispatched',
 'to',
 'the',
 'moon',
 'Pandora',
 'on',
 'a',
 'unique',
 'mission,',
 'but',
 'becomes',
 'torn',
 'between',
 'following',
 'orders',
 'and',
 'protecting',
 'an',
 'alien',
 'civilization.',
 'Action',
 'Adventure',
 'Fantasy',
 'ScienceFiction',
 'cultureclash',
 'future',
 'spacewar',
 'spacecolony',
 'society',
 'spacetravel',
 'futuristic',
 'romance',
 'space',
 'alien',
 'tribe',
 'alienplanet',
 'cgi',
 'marine',
 'soldier',
 'battle',
 'loveaffair',
 'antiwar',
 'powerrelations',
 'mindandsoul',
 '3d',
 'SamWorthington',
 'ZoeSaldana',
 'SigourneyWeaver',
 'StephenLang',
 'MichelleRodriguez',
 'GiovanniRibisi',
 'JoelDavidMoore',
 'CCHPounder',
 'WesStudi',
 'LazAlonso',
 'DileepRao',
 'MattGerald',
 'SeanAnthonyMoran',
 'JasonWhyte',
 'ScottLawrence',
 'KellyKilgour',
 'JamesPatrickPitt',
 'SeanPatrickMurphy',
 'PeterDillon',
 'KevinDorman',
 'KelsonHenderson',
 'DavidVanHorn',
 'J

The join function is used in this case to concatenate the elements of a list into a single string. This is necessary because the dataframe we are working with has a column called 'tags' that contains lists of strings, rather than just strings.

In [68]:
new_df["tags"] = new_df["tags"].apply(lambda x : " ".join (x))

C:\Users\haric\AppData\Local\Temp\ipykernel_1908\3526719444.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["tags"] = new_df["tags"].apply(lambda x : " ".join (x))


In [69]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

In [70]:
new_df["tags"] = new_df["tags"].apply(lambda x : x.lower())

C:\Users\haric\AppData\Local\Temp\ipykernel_1908\1044876906.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["tags"] = new_df["tags"].apply(lambda x : x.lower())


In [71]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  in the 22nd century, a paraplegic marine is di...  
1  captain barbossa, long believed to be dead, ha...  
2  a cryptic message from bond’s past sends him o...  
3  following the death of district attorney harve...  
4  john carter is a war-weary, former military ca...

In [87]:
new_df["tags"][0]

'in the 22nd century, a paraplegic marine is dispatched to the moon pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. action adventure fantasy sciencefiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d samworthington zoesaldana sigourneyweaver stephenlang michellerodriguez giovanniribisi joeldavidmoore cchpounder wesstudi lazalonso dileeprao mattgerald seananthonymoran jasonwhyte scottlawrence kellykilgour jamespatrickpitt seanpatrickmurphy peterdillon kevindorman kelsonhenderson davidvanhorn jacobtomuri michaelblain-rozgay joncurry lukehawker woodyschultz petermensah soniayee jahnelcurfman ilramchoi kylawarren lisaroumain debrawilson chrismala taylorkibby jodielandau julielamm cullenb.madden josephbradymadden frankietorres austinwilson sarawilson tamicawashington-miller lucybriant nathanm

from here on words we came to new concept that is called vectorization we dont have a good idea about it  but still we can do it  where here i will convert all the tags into vectors and i will get most frequent words from the tags for the particuar movie  and get vector rom it and i will place it in the graph and i will recommend the movies which is very near to my vecotor

1 for this we should install the package sciketlearn we need to get the function count vectorizer which  is  very useful

In a bag-of-words model, which is often used in text processing tasks like document classification, we typically want to ignore rare words and focus on the most frequent ones. This is because the most frequent words often carry more information than the less frequent ones.

The max_features parameter in the CountVectorizer constructor specifies the maximum number of features (words) to keep. By setting max_features = 5000, we are keeping the top 5000 most frequent words in the dataset. This can help in reducing the dimensionality of the dataset and improving computational efficiency.

Additionally, the stop_words = 'english' parameter in the CountVectorizer constructor specifies a predefined set of English stop words that should be ignored. Stop words are common words that don't carry much meaning, such as "and", "the", "in", etc.

By using max_features and stop_words, we can improve the performance and efficiency of the CountVectorizer and make it less sensitive to irrelevant words.

In [74]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 5000,stop_words = 'english')

In [76]:
vectors = cv.fit_transform(new_df["tags"]).toarray()

In [80]:
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [81]:

vectors.shape

(4806, 5000)

In [82]:
vectors[0]

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

To retrieve the feature names from the CountVectorizer, you can use the get_feature_names() method.

Here is an example of how to use this method:

python
Copy code
from sklearn.feature_extraction.text import CountVectorizer

# Sample text data
corpus = ['This is the first document.',
          'This document is the second document.',
          'And this is the third one.',
          'Is this the first document?']

# Create a CountVectorizer instance
cv = CountVectorizer(max_features = 5000,stop_words = 'english')

# Fit the CountVectorizer and transform the text data into a matrix of token counts
X = cv.fit_transform(corpus)

# Retrieve the feature names
feature_names = cv.get_feature_names()

# Print the feature names
print(feature_names)
In this example, the get_feature_names() method is used to retrieve the feature names from the fitted CountVectorizer. These feature names represent the words that were tokenized from the text data.

Remember, it is important to use the fit_transform method first, as the CountVectorizer must be fitted to the data before you can retrieve the feature names.

In [85]:
cv.get_feature_names()

['000',
 '007',
 '10',
 '100',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '18',
 '18th',
 '19',
 '1930s',
 '1940s',
 '1950s',
 '1960s',
 '1970s',
 '1980s',
 '19th',
 '19thcentury',
 '20',
 '20th',
 '24',
 '25',
 '30',
 '3d',
 '40',
 '50',
 '60',
 '60s',
 'aaron',
 'aaroneckhart',
 'aarontaylor',
 'aasifmandvi',
 'abandoned',
 'abducted',
 'abigailbreslin',
 'ability',
 'able',
 'abuse',
 'abusive',
 'academy',
 'accept',
 'accepts',
 'access',
 'accident',
 'accidentally',
 'accompanied',
 'account',
 'accountant',
 'accused',
 'ace',
 'achieve',
 'act',
 'action',
 'actions',
 'activist',
 'activities',
 'actor',
 'actors',
 'actress',
 'acts',
 'actual',
 'actually',
 'adam',
 'adambrody',
 'adamgoldberg',
 'adamlefevre',
 'adams',
 'adamsandler',
 'adamscott',
 'adamshankman',
 'adaptation',
 'addict',
 'addicted',
 'addiction',
 'adewaleakinnuoye',
 'adolescence',
 'adopted',
 'adoption',
 'adrianmartinez',
 'adrienbrody',
 'adult',
 'adultery',
 'adulthood',
 'adults',
 'ad

here we have to do the stemming beacues when we see the vectors there are very similar wods which contains the same meaning to achiesve that we need to us it

In [86]:
import nltk
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer( )

In [89]:
def stem(text):
    y = [ ]
    for i in text.split():
        y.append(ps.stem(i))
    return " ".join(y)

In [90]:
stem("in the 22nd century, a paraplegic marine is dispatched to the moon pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. action adventure fantasy sciencefiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d samworthington zoesaldana sigourneyweaver stephenlang michellerodriguez giovanniribisi joeldavidmoore cchpounder wesstudi lazalonso dileeprao mattgerald seananthonymoran jasonwhyte scottlawrence kellykilgour jamespatrickpitt seanpatrickmurphy peterdillon kevindorman kelsonhenderson davidvanhorn jacobtomuri michaelblain-rozgay joncurry lukehawker woodyschultz petermensah soniayee jahnelcurfman ilramchoi kylawarren lisaroumain debrawilson chrismala taylorkibby jodielandau julielamm cullenb.madden josephbradymadden frankietorres austinwilson sarawilson tamicawashington-miller lucybriant nathanmeister gerryblair matthewchamberlain paulyates wraywilson jamesgaylyn melvinlenoclarkiii carvonfutrell brandonjelkes micahmoch hanniyahmuhammad christophernolen christaoliver aprilmariethomas bravitaa.threatt colinbleasdale mikebodnar mattclayton nicoledionne jamieharrison allanhenry anthonyingruber ashleyjeffery deanknowsley josephmika-hunt terrynotary kaipantano loganpithyou stuartpollock raja garethruck rhiansheehan t.j.storm jodietaylor aliciavela-bailey richardwhiteside nikiezambo julenerenee jamescameron'")

"in the 22nd century, a parapleg marin is dispatch to the moon pandora on a uniqu mission, but becom torn between follow order and protect an alien civilization. action adventur fantasi sciencefict cultureclash futur spacewar spacecoloni societi spacetravel futurist romanc space alien tribe alienplanet cgi marin soldier battl loveaffair antiwar powerrel mindandsoul 3d samworthington zoesaldana sigourneyweav stephenlang michellerodriguez giovanniribisi joeldavidmoor cchpounder wesstudi lazalonso dileeprao mattgerald seananthonymoran jasonwhyt scottlawr kellykilgour jamespatrickpitt seanpatrickmurphi peterdillon kevindorman kelsonhenderson davidvanhorn jacobtomuri michaelblain-rozgay joncurri lukehawk woodyschultz petermensah soniaye jahnelcurfman ilramchoi kylawarren lisaroumain debrawilson chrismala taylorkibbi jodielandau julielamm cullenb.madden josephbradymadden frankietorr austinwilson sarawilson tamicawashington-mil lucybri nathanmeist gerryblair matthewchamberlain paulyat wraywil

In [91]:
new_df["tags"] = new_df["tags"].apply(stem)

C:\Users\haric\AppData\Local\Temp\ipykernel_1908\825579503.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["tags"] = new_df["tags"].apply(stem)


In [92]:
new_df["tags"][1]

"captain barbossa, long believ to be dead, ha come back to life and is head to the edg of the earth with will turner and elizabeth swann. but noth is quit as it seems. adventur fantasi action ocean drugabus exoticisland eastindiatradingcompani loveofone'slif traitor shipwreck strongwoman ship allianc calypso afterlif fighter pirat swashbuckl aftercreditssting johnnydepp orlandobloom keiraknightley stellanskarsgård chowyun-fat billnighi geoffreyrush jackdavenport kevinmcn tomholland naomieharri jonathanpryc keithrichard leearenberg mackenziecrook gregelli davidbaili martinklebba davidschofield laurenmah vanessabranch angusbarnett gilesnew reggiele dominicscottkay takayofisch davidmeuni ho-kwants andybeckwith peterdonaldbadalamentiii christophers.capp keithrichard hakeemkae-kazim ghassanmassoud goreverbinski"

now we got every thing which we need like vectors and all now we have 4806 movies and 5000 words in each movie 
as discussed we have to calculate the distance between the vectors which are very similar now what we have to do is we need to calculate the angle between them{(cosine distance(which  means if if the distance between two vectors then it is a same movie i hope you get it HARIIIIII))} not the point to point distance(euclidean distance) we have 

and distance is inversly proportinal to similarity

In [93]:
from sklearn.metrics.pairwise import cosine_similarity

In [94]:
similarity = cosine_similarity(vectors)

In [95]:
similarity[0]

array([1.        , 0.07142857, 0.05216405, ..., 0.02326211, 0.02571722,
       0.        ])

In [105]:
list(enumerate(similarity[0]))

[(0, 1.0000000000000002),
 (1, 0.07142857142857144),
 (2, 0.052164053095730106),
 (3, 0.03308611113236434),
 (4, 0.16101529717988267),
 (5, 0.09156219276643894),
 (6, 0.020088585092257697),
 (7, 0.13829563644637993),
 (8, 0.04761904761904762),
 (9, 0.05878963512522154),
 (10, 0.10286889997472794),
 (11, 0.07229440390678199),
 (12, 0.07715167498104596),
 (13, 0.039513038984679984),
 (14, 0.09412416106700235),
 (15, 0.04501491366063437),
 (16, 0.06482037235521643),
 (17, 0.130031582934249),
 (18, 0.10044292546128848),
 (19, 0.07655590023509529),
 (20, 0.05111986324432477),
 (21, 0.07715167498104596),
 (22, 0.05832118435198043),
 (23, 0.07824607964359516),
 (24, 0.04941662111074008),
 (25, 0.032712244767539374),
 (26, 0.13226001425322162),
 (27, 0.1649572197684645),
 (28, 0.10585122480499262),
 (29, 0.06419407387663695),
 (30, 0.05832118435198042),
 (31, 0.13041013273932528),
 (32, 0.07715167498104596),
 (33, 0.07715167498104596),
 (34, 0.0),
 (35, 0.07938841860374446),
 (36, 0.1860968420

here we got he similarity of  every movie and in the first place above you can see it is 1 which means it is comaparing its own movie so it is 1.

In [103]:
def recomender(movie):
    movie_index = new_df[new_df["title"]== movie].index[0]
    return movie_index

recomender("The Dark Knight Rises")

3

here we have to use the enumumerate function because when we gave the move name we are able to get the  similarity but we want to sort in a way where we want top 5 if we sort that the indexes of similarities  would be changed  so that we should use the enumurate function

and also at first we are  getting the similaities of the movies by giving the index of the movie in our df so that it go to the above loc where we get all the similaiities of the movie with other movies 

In [126]:
sorted(list(enumerate(similarity[0])),reverse=True,key = lambda x : x[1])
#this lambda function is to sort

AttributeError: 'list' object has no attribute 'astype'

In [113]:
sorted(list(enumerate(similarity[0])),reverse=True,key = lambda x : x[1])[0:6]
#print(top_5)excluding the 0 index because that is our movie

[(1920, 0.23473823893078552),
 (1216, 0.23294541397390256),
 (582, 0.2309782890611944),
 (539, 0.2252817784447915),
 (507, 0.21912524504463887)]

In [134]:
def recomender(movie):
    movie_index = new_df[new_df["title"]== movie].index[0]
    distances = similarity[movie_index]
    movie_list = sorted(list(enumerate(distances)),reverse=True,key = lambda x : x[1])[0:6]
    for i in movie_list:
        print(new_df.iloc[i[0]]["title"])

recomender("Avatar")

Avatar
Lifeforce
Aliens vs Predator: Requiem
Battle: Los Angeles
Titan A.E.
Independence Day


here i[0] because  we have a key value pair so if we give i[0] then it will give the x value in(x,y)where x is our index and bades on that we are getting the values from the new_df

In [127]:
import pickle

In [131]:
pickle.dump(new_df.to_dict,open('movie_dict.pkl','wb'))

In [133]:
pickle.dump(similarity,open('similarity.pkl','wb'))